File: src/ios/HivePlugin.swift
/*
 * Copyright (c) 2019 Elastos Foundation
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import Foundation
import ElastosHiveSDK

@objc(HivePlugin)
class HivePlugin : TrinityPlugin {
    internal static let LOGIN : Int = 1
    internal static let RESULT: Int = 2

    internal var loginCallbackId:  String = ""
    internal var resultCallbackId: String = ""

    internal var count: Int = 1

    var callbackId: String = ""

    @objc func initVal(_ command: CDVInvokedUrlCommand) {
    }

    @objc func success(_ command: CDVInvokedUrlCommand, retAsString: String) {
        let result = CDVPluginResult(status: CDVCommandStatus_OK,
                                     messageAs: retAsString);

        self.commandDelegate.send(result, callbackId: command.callbackId)
    }

    @objc func success(_ command: CDVInvokedUrlCommand, retAsDict: NSDictionary) {
        let result = CDVPluginResult(status: CDVCommandStatus_OK,
                                     messageAs: (retAsDict as! [AnyHashable : Any]));

        self.commandDelegate.send(result, callbackId: command.callbackId)
    }

    @objc func error(_ command: CDVInvokedUrlCommand, retAsString: String) {
        let result = CDVPluginResult(status: CDVCommandStatus_ERROR,
                                     messageAs: retAsString);

        self.commandDelegate.send(result, callbackId: command.callbackId)
    }

    @objc func getVersion(_ command: CDVInvokedUrlCommand) {
        self.success(command, retAsString: "HiveSDK-v0.0.1");
    }

    @objc func setListener(_ command: CDVInvokedUrlCommand) {
        let type = command.arguments[0] as? Int ?? 0

        switch (type) {
        case HivePlugin.LOGIN:
            loginCallbackId = command.callbackId

        case HivePlugin.RESULT:
            resultCallbackId = command.callbackId

        default:
            self.error(command, retAsString: "Expected one non-empty let argument.")
        }

        // Don't return any result now
        let result = CDVPluginResult(status: CDVCommandStatus_NO_RESULT);
        result?.setKeepCallbackAs(true);
        self.commandDelegate.send(result, callbackId: command.callbackId)
    }

    @objc func createClient(_ command: CDVInvokedUrlCommand) {
        let dataDir = command.arguments[0] as? String ?? ""
        let options = command.arguments[1] as? String ?? ""

        let map: ObjectMap = ObjectMap<HiveClientHandle>.acquire(1)
        let client: HiveClientHandle? = ClientBuilder.createClient(dataDir, options)
        guard client != nil else {
            self.error(command, retAsString: "Client create error")
            return
        }

        let objId: Int = count
        ObjectMap.asClientMap(map).put(objId, client!)
        count += 1

        let ret: NSDictionary = [ "id": objId ]
        self.success(command, retAsDict: ret);
    }

    @objc func login(_ command: CDVInvokedUrlCommand) {
        let mapId = command.arguments[0] as? Int ?? 0
        let objId = command.arguments[1] as? Int ?? 0
        let hdrId = command.arguments[2] as? Int ?? 0

        let map: ObjectMap = ObjectMap<HiveClientHandle>.acquire(mapId)
        guard ObjectMap.isClientMap(map) else {
            return
        }

        let backgroundQueue = DispatchQueue(label: "org.elastos.hive.queue",
                                            qos: .background, target: nil)
        backgroundQueue.async {
            var ret = Dictionary<String, Any>()

            _ = try! ObjectMap.asClientMap(map).get(objId).login(
                LoginHandler(hdrId, self.loginCallbackId, self.commandDelegate)
            )

            ret["result"] = "success"
            self.success(command, retAsDict: ret as NSDictionary);
        }
    }

    @objc func logout(_ command: CDVInvokedUrlCommand) {
        let mapId = command.arguments[0] as? Int ?? 0
        let objId = command.arguments[1] as? Int ?? 0

        let map: ObjectMap = ObjectMap<HiveClientHandle>.acquire(mapId)
        guard ObjectMap.isClientMap(map) else {
            return
        }

        let backgroundQueue = DispatchQueue(label: "org.elastos.hive.queue",
                                            qos: .background, target: nil)
        backgroundQueue.async {
            var ret = Dictionary<String, Any>()

            _ = try! ObjectMap.asClientMap(map).get(objId).logout()

            ret["result"] = "success"
            self.success(command, retAsDict: ret as NSDictionary);
        }
    }

    @objc func getLastInfo(_ command: CDVInvokedUrlCommand) {
        let mapId = command.arguments[0] as? Int ?? 0
        let objId = command.arguments[1] as? Int ?? 0

        let map: ObjectMap = ObjectMap<HiveClientHandle>.acquire(mapId)
        if (ObjectMap.isClientMap(map)) {
            var holder: DictionaryHolder<HiveClientInfo>
            var info: HiveClientInfo?

            info = ObjectMap.asClientMap(map).get(objId).lastInfo
            if (info == nil) {
                self.error(command, retAsString: "Expected one non-empty let argument.")
                return
            }

            holder = DictionaryHolder<HiveClientInfo>(info!)
            _ = holder.put(HiveClientInfo.userId)
                .put(HiveClientInfo.name)
                .put(HiveClientInfo.email)
                .put(HiveClientInfo.phoneNo)
                .put(HiveClientInfo.region)

            self.success(command, retAsDict: holder.get() as NSDictionary)
            return;
        }

        if (ObjectMap.isDriveMap(map)) {
            var holder: DictionaryHolder<HiveDriveInfo>
            var info: HiveDriveInfo

            info = ObjectMap.asDriveMap(map).get(objId).lastInfo
            holder = DictionaryHolder<HiveDriveInfo>(info)
            _ = holder.put(HiveDriveInfo.driveId)

            self.success(command, retAsDict: holder.get() as NSDictionary)
            return;
        }

        if (ObjectMap.isDirMap(map)) {
            var holder: DictionaryHolder<HiveDirectoryInfo>
            var info: HiveDirectoryInfo

            info = ObjectMap.asDirMap(map).get(objId).lastInfo!
            holder = DictionaryHolder<HiveDirectoryInfo>(info)
            _ =  holder.put(HiveDirectoryInfo.itemId)
                .put(HiveDirectoryInfo.name)
                .put(HiveDirectoryInfo.childCount)

            self.success(command, retAsDict: holder.get() as NSDictionary)
            return;
        }

        if (ObjectMap.isFileMap(map)) {
            var holder: DictionaryHolder<HiveFileInfo>
            var info: HiveFileInfo

            info = ObjectMap.asFileMap(map).get(objId).lastInfo!
            holder = DictionaryHolder<HiveFileInfo>(info)
            _ = holder.put(HiveFileInfo.itemId)
                .put(HiveFileInfo.name)
                .put(HiveFileInfo.size)

            self.success(command, retAsDict: holder.get() as NSDictionary)
            return;
        }
    }

    @objc func getInfo(_ command: CDVInvokedUrlCommand) {
        let mapId = command.arguments[0] as? Int ?? 0
        let objId = command.arguments[1] as? Int ?? 0
        let hdrId = command.arguments[2] as? Int ?? 0

        let map: ObjectMap = ObjectMap<HiveClientHandle>.acquire(mapId)
        if (ObjectMap.isClientMap(map)) {
            _ = ObjectMap.asClientMap(map).get(objId).lastUpdatedInfo(
                handleBy: ResultHandler<HiveClientInfo>(hdrId, self.resultCallbackId, self.commandDelegate)
            )
            return
        }

        if (ObjectMap.isDriveMap(map)) {
            _ = ObjectMap.asDriveMap(map).get(objId).lastUpdatedInfo(
                handleBy: ResultHandler<HiveDriveInfo>(hdrId, self.resultCallbackId, self.commandDelegate)
            )
            return
        }

        if (ObjectMap.isDirMap(map)) {
            _ = ObjectMap.asDirMap(map).get(objId).lastUpdatedInfo(
                handleBy: ResultHandler<HiveDirectoryInfo>(hdrId, self.resultCallbackId, self.commandDelegate)
            )
            return
        }

        if (ObjectMap.isFileMap(map)) {
            _ = ObjectMap.asFileMap(map).get(objId).lastUpdatedInfo(
                handleBy: ResultHandler<HiveFileInfo>(hdrId, self.resultCallbackId, self.commandDelegate)
            )
        }
    }

    @objc func getDefDrive(_ command: CDVInvokedUrlCommand) {
        let mapId = command.arguments[0] as? Int ?? 0
        let objId = command.arguments[1] as? Int ?? 0
        let hdrId = command.arguments[2] as? Int ?? 0

        let map: ObjectMap = ObjectMap<HiveClientHandle>.acquire(mapId)
        if (ObjectMap.isClientMap(map)) {
            _ = ObjectMap.asClientMap(map).get(objId).defaultDriveHandle(
                handleBy: ResultHandler<HiveDriveHandle>(hdrId, self.resultCallbackId, self.commandDelegate)
            )
            return
        }
    }

    @objc func rootDir(_ command: CDVInvokedUrlCommand) {
        let mapId = command.arguments[0] as? Int ?? 0
        let objId = command.arguments[1] as? Int ?? 0
        let hdrId = command.arguments[2] as? Int ?? 0

        let map: ObjectMap = ObjectMap<HiveClientHandle>.acquire(mapId)
        if (ObjectMap.isDriveMap(map)) {
            _ = ObjectMap.asDriveMap(map).get(objId).rootDirectoryHandle(
                handleBy: ResultHandler<HiveDirectoryHandle>(hdrId, self.resultCallbackId, self.commandDelegate)
            )
            return
        }
    }

    @objc func createDir(_ command: CDVInvokedUrlCommand) {
        let mapId = command.arguments[0] as? Int ?? 0
        let objId = command.arguments[1] as? Int ?? 0
        let hdrId = command.arguments[2] as? Int ?? 0
        let path  = command.arguments[3] as? String ?? ""

        let map: ObjectMap = ObjectMap<HiveClientHandle>.acquire(mapId)
        if (ObjectMap.isDriveMap(map)) {
            _ = ObjectMap.asDriveMap(map).get(objId).createDirectory(withPath: path,
                handleBy: ResultHandler<HiveDirectoryHandle>(hdrId, self.resultCallbackId, self.commandDelegate)
            )
            return
        }
        if (ObjectMap.isDirMap(map)) {
            _ = ObjectMap.asDirMap(map).get(objId).createDirectory(withName: path,
                handleBy: ResultHandler<HiveDirectoryHandle>(hdrId, self.resultCallbackId, self.commandDelegate)
            )
            return
        }
    }

    @objc func getDir(_ command: CDVInvokedUrlCommand) {
        let mapId = command.arguments[0] as? Int ?? 0
        let objId = command.arguments[1] as? Int ?? 0
        let hdrId = command.arguments[2] as? Int ?? 0
        let  path = command.arguments[3] as? String ?? ""

        let map: ObjectMap = ObjectMap<HiveClientHandle>.acquire(mapId)
        if (ObjectMap.isDriveMap(map)) {
            _ = ObjectMap.asDriveMap(map).get(objId).directoryHandle(atPath: path,
                handleBy: ResultHandler<HiveDirectoryHandle>(hdrId, self.resultCallbackId, self.commandDelegate)
            )
            return
        }
        if (ObjectMap.isDirMap(map)) {
            _ = ObjectMap.asDirMap(map).get(objId).directoryHandle(atName: path,
                handleBy: ResultHandler<HiveDirectoryHandle>(hdrId, self.resultCallbackId, self.commandDelegate)
            )
            return
        }
    }

    @objc func createFile(_ command: CDVInvokedUrlCommand) {
        let mapId = command.arguments[0] as? Int ?? 0
        let objId = command.arguments[1] as? Int ?? 0
        let hdrId = command.arguments[2] as? Int ?? 0
        let  path = command.arguments[3] as? String ?? ""

        let map: ObjectMap = ObjectMap<HiveClientHandle>.acquire(mapId)
        if (ObjectMap.isDriveMap(map)) {
            _ = ObjectMap.asDriveMap(map).get(objId).createFile(withPath: path,
                handleBy: ResultHandler<HiveFileHandle>(hdrId, self.resultCallbackId, self.commandDelegate)
            )
            return
        }
        if (ObjectMap.isDirMap(map)) {
            _ = ObjectMap.asDirMap(map).get(objId).createFile(withName: path,
                handleBy: ResultHandler<HiveFileHandle>(hdrId, self.resultCallbackId, self.commandDelegate)
            )
            return
        }
    }

    @objc func getFile(_ command: CDVInvokedUrlCommand) {
        let mapId = command.arguments[0] as? Int ?? 0
        let objId = command.arguments[1] as? Int ?? 0
        let hdrId = command.arguments[2] as? Int ?? 0
        let  path = command.arguments[3] as? String ?? ""

        let map: ObjectMap = ObjectMap<HiveClientHandle>.acquire(mapId)
        if (ObjectMap.isDriveMap(map)) {
            _ = ObjectMap.asDriveMap(map).get(objId).fileHandle(atPath: path,
                handleBy: ResultHandler<HiveFileHandle>(hdrId, self.resultCallbackId, self.commandDelegate)
            )
            return
        }
        if (ObjectMap.isDirMap(map)) {
            _ = ObjectMap.asDirMap(map).get(objId).fileHandle(atName: path,
                handleBy: ResultHandler<HiveFileHandle>(hdrId, self.resultCallbackId, self.commandDelegate)
            )
            return
        }
    }

    @objc func getItemInfo(_ command: CDVInvokedUrlCommand) {
        let mapId = command.arguments[0] as? Int ?? 0
        let objId = command.arguments[1] as? Int ?? 0
        let hdrId = command.arguments[2] as? Int ?? 0
        let  path = command.arguments[3] as? String ?? ""

        let map: ObjectMap = ObjectMap<HiveClientHandle>.acquire(mapId)
        if (ObjectMap.isDriveMap(map)) {
            _ = ObjectMap.asDriveMap(map).get(objId).getItemInfo(path,
                handleBy: ResultHandler<HiveItemInfo>(hdrId, self.resultCallbackId, self.commandDelegate)
            )
            return
        }
    }

    @objc func getChildren(_ command: CDVInvokedUrlCommand) {
        let mapId = command.arguments[0] as? Int ?? 0
        let objId = command.arguments[1] as? Int ?? 0
        let hdrId = command.arguments[2] as? Int ?? 0

        let map: ObjectMap = ObjectMap<HiveClientHandle>.acquire(mapId)
        if (ObjectMap.isDirMap(map)) {
            _ = ObjectMap.asDirMap(map).get(objId).getChildren(
                handleBy: ResultHandler<HiveChildren>(hdrId, self.resultCallbackId, self.commandDelegate)
            )
            return
        }
    }

    @objc func moveTo(_ command: CDVInvokedUrlCommand) {
        let mapId = command.arguments[0] as? Int ?? 0
        let objId = command.arguments[1] as? Int ?? 0
        let hdrId = command.arguments[2] as? Int ?? 0
        let  path = command.arguments[3] as? String ?? ""

        let map: ObjectMap = ObjectMap<HiveClientHandle>.acquire(mapId)
        if (ObjectMap.isDirMap(map)) {
            _ = ObjectMap.asDirMap(map).get(objId).moveTo(newPath: path,
                handleBy: ResultHandler<Void>(hdrId, self.resultCallbackId, self.commandDelegate)
            )
            return
        }
        if (ObjectMap.isFileMap(map)) {
            _ = ObjectMap.asFileMap(map).get(objId).moveTo(newPath: path,
                handleBy: ResultHandler<Void>(hdrId, self.resultCallbackId, self.commandDelegate)
            )
            return
        }
    }

    @objc func copyTo(_ command: CDVInvokedUrlCommand) {
        let mapId = command.arguments[0] as? Int ?? 0
        let objId = command.arguments[1] as? Int ?? 0
        let hdrId = command.arguments[2] as? Int ?? 0
        let  path = command.arguments[3] as? String ?? ""

        let map: ObjectMap = ObjectMap<HiveClientHandle>.acquire(mapId)
        if (ObjectMap.isDirMap(map)) {
            _ = ObjectMap.asDirMap(map).get(objId).copyTo(newPath: path,
                handleBy: ResultHandler<Void>(hdrId, self.resultCallbackId, self.commandDelegate)
            )
            return
        }
        if (ObjectMap.isFileMap(map)) {
            _ = ObjectMap.asFileMap(map).get(objId).copyTo(newPath: path,
                handleBy: ResultHandler<Void>(hdrId, self.resultCallbackId, self.commandDelegate)
            )
            return
        }
    }

    @objc func deleteItem(_ command: CDVInvokedUrlCommand) {
        let mapId = command.arguments[0] as? Int ?? 0
        let objId = command.arguments[1] as? Int ?? 0
        let hdrId = command.arguments[2] as? Int ?? 0

        let map: ObjectMap = ObjectMap<HiveClientHandle>.acquire(mapId)
        if (ObjectMap.isDirMap(map)) {
            _ = ObjectMap.asDirMap(map).get(objId).deleteItem(
                handleBy: ResultHandler<Void>(hdrId, self.resultCallbackId, self.commandDelegate)
            )
            return
        }
        if (ObjectMap.isFileMap(map)) {
            _ = ObjectMap.asFileMap(map).get(objId).deleteItem(
                handleBy: ResultHandler<Void>(hdrId, self.resultCallbackId, self.commandDelegate)
            )
            return
        }
    }
}

File: src/android/HivePlugin.java
/*
 * Copyright (c) 2019 Elastos Foundation
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

  package org.elastos.trinity.plugins.hive;

  import android.util.Base64;

  import org.apache.cordova.CordovaPlugin;
  import org.apache.cordova.CallbackContext;

  import org.apache.cordova.PluginResult;
  import org.elastos.trinity.runtime.TrinityPlugin;
  import org.json.JSONArray;
  import org.json.JSONException;
  import org.json.JSONObject;

  import java.util.List;
  import java.util.Map;
  import java.util.HashMap;

  import org.elastos.hive.*;
  import org.elastos.hive.Void;
  import org.elastos.hive.File;
  import org.elastos.hive.HiveException;

  /**
   * This class echoes a string called from JavaScript.
   */
  public class HivePlugin extends TrinityPlugin {
      private static final int LOGIN  = 1;
      private static final int RESULT = 2;

      private CallbackContext loginCallbackCtxt  = null;
      private CallbackContext resultCallbackCtxt = null;

      @Override
      public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
          try {
              switch(action) {
              case "getVersion":
                  this.getVersion(args, callbackContext);
                  break;

              case "setListener":
                  this.setListener(args, callbackContext);
                  break;

              case "createClient":
                  this.createClient(args, callbackContext);
                  break;

              case "login":
                  this.login(args, callbackContext);
                  break;

              case "logout":
                  this.logout(args, callbackContext);
                  break;

              case "getLastInfo":
                  this.getLastInfo(args, callbackContext);
                  break;

              case "getInfo":
                  this.getInfo(args, callbackContext);
                  break;

              case "getDefDrive":
                  this.getDefDrive(args, callbackContext);
                  break;

              case "rootDir":
                  this.rootDirectory(args, callbackContext);
                  break;

              case "createDir":
                  this.createDirectory(args, callbackContext);
                  break;

              case "getDir":
                  this.getDirectory(args, callbackContext);
                  break;

              case "createFile":
                  this.createFile(args, callbackContext);
                  break;

              case "getFile":
                  this.getFile(args, callbackContext);
                  break;

              case "getItemInfo":
                  this.getItemInfo(args, callbackContext);
                  break;

              case "getChildren":
                  this.getChildren(args, callbackContext);
                  break;

              case "moveTo":
                  this.moveTo(args, callbackContext);
                  break;

              case "copyTo":
                  this.moveTo(args, callbackContext);
                  break;

              case "deleteItem":
                  this.deleteItem(args, callbackContext);
                  break;

              default:
                  return false;
              }
          } catch (JSONException e) {
              //e.printStack();
          }
          return true;
      }

      private void getVersion(JSONArray args, CallbackContext callbackContext) throws JSONException {
          String version = "ElastosHiveSDK-v0.1";
          callbackContext.success(version);
      }

      private void setListener(JSONArray args, CallbackContext callbackContext) throws JSONException {
          Integer type = args.getInt(0);

          switch(type) {
          case LOGIN:
              loginCallbackCtxt = callbackContext;
              break;

          case RESULT:
              resultCallbackCtxt = callbackContext;
              break;
          }

          PluginResult result = new PluginResult(PluginResult.Status.NO_RESULT);
          result.setKeepCallback(true);
          callbackContext.sendPluginResult(result);
      }

      private void createClient(JSONArray args, CallbackContext callbackContext) throws JSONException {
          String dataDir = args.getString(0);
          String options = args.getString(1);

          java.io.File dirFile = new java.io.File(cordova.getActivity().getFilesDir() + "/data/hive" + dataDir);
          if (!dirFile.exists())
               dirFile.mkdirs();

          ObjectMap map = ObjectMap.acquire(1);
          Client client = ClientBuilder.create(dataDir, options);
          Integer objId = System.identityHashCode(client);
          ObjectMap.toClientMap(map).put(objId, client);

          JSONObject ret= new JSONObject();
          ret.put("id", objId);
          callbackContext.success(ret);
      }

      private void login(JSONArray args, CallbackContext callbackContext) throws JSONException {
          Integer mapId = args.getInt(0);
          Integer objId = args.getInt(1);
          int handlerId = args.getInt(2);

          ObjectMap map = ObjectMap.acquire(mapId);
          if (!ObjectMap.isClientMap(map))
              return;

          JSONObject ret = new JSONObject();
          new Thread(() -> {
              try {
                  ObjectMap.toClientMap(map).get(objId).login(
                      new LoginHandler(handlerId, loginCallbackCtxt)
                  );

                  ret.put("result", "success");
              } catch(JSONException e) {
                  callbackContext.error("error");
              } catch(HiveException e) {
                  callbackContext.error("error");
              }
          }).start();
          callbackContext.success(ret);
      }

      private void logout(JSONArray args, CallbackContext callbackContext) throws JSONException {
          Integer mapId = args.getInt(0);
          Integer objId = args.getInt(1);

          ObjectMap map = ObjectMap.acquire(mapId);
          if (!ObjectMap.isClientMap(map))
              return;

          new Thread(() -> {
              try {
                  ObjectMap.toClientMap(map).get(objId).logout();

                  JSONObject ret = new JSONObject();
                  ret.put("result", "success");
                  callbackContext.success(ret);
              } catch (JSONException e) {
                  callbackContext.error("error");
              } catch (HiveException e) {
                  callbackContext.error("error");
              }
          }).start();
      }

      private void getLastInfo(JSONArray args, CallbackContext callbackContext) throws JSONException {
          Integer mapId = args.getInt(0);
          Integer objId = args.getInt(1);

          ObjectMap map = ObjectMap.acquire(mapId);
          if (ObjectMap.isClientMap(map)) {
              JSONObjectHolder<Client.Info> holder;
              Client.Info info;

              info = ObjectMap.toClientMap(map).get(objId).getLastInfo();
              if (info == null) {
                  callbackContext.error("no info");
                  return;
              }

              holder = new JSONObjectHolder<Client.Info>(info);
              holder.put(Client.Info.userId)
                    .put(Client.Info.name)
                    .put(Client.Info.email)
                    .put(Client.Info.phoneNo)
                    .put(Client.Info.region);

              callbackContext.success(holder.get());
              return;
          }

          if (ObjectMap.isDriveMap(map)) {
              JSONObjectHolder<Drive.Info> holder;
              Drive.Info info;

              info = ObjectMap.toDriveMap(map).get(objId).getLastInfo();
              holder = new JSONObjectHolder<Drive.Info>(info);
              holder.put(Drive.Info.driveId);

              callbackContext.success(holder.get());
              return;
          }

          if (ObjectMap.isDirMap(map)) {
              JSONObjectHolder<Directory.Info> holder;
              Directory.Info info;

              info = ObjectMap.toDirMap(map).get(objId).getLastInfo();
              holder = new JSONObjectHolder<Directory.Info>(info);
              holder.put(Directory.Info.itemId)
                    .put(Directory.Info.name)
                    .put(Directory.Info.childCount);

              callbackContext.success(holder.get());
              return;
          }

          if (ObjectMap.isFileMap(map)) {
              JSONObjectHolder<File.Info> holder;
              File.Info info;

              info = ObjectMap.toFileMap(map).get(objId).getLastInfo();
              holder = new JSONObjectHolder<File.Info>(info);
              holder.put(File.Info.itemId)
                    .put(File.Info.name)
                    .put(File.Info.size);

              callbackContext.success(holder.get());
              return;
          }
      }

      private void getInfo(JSONArray args, CallbackContext callbackContext) throws JSONException {
          Integer mapId = args.getInt(0);
          Integer objId = args.getInt(1);
          int handlerId = args.getInt(2);

          ObjectMap map = ObjectMap.acquire(mapId);
          if (ObjectMap.isClientMap(map)) {
              ObjectMap.toClientMap(map).get(objId).getInfo(
                  new ResultHandler<Client.Info>(handlerId, resultCallbackCtxt)
              );
              return;
          }

          if (ObjectMap.isDriveMap(map)) {
              ObjectMap.toDriveMap(map).get(objId).getInfo(
                  new ResultHandler<Drive.Info>(handlerId, resultCallbackCtxt)
              );
              return;
          }

          if (ObjectMap.isDirMap(map)) {
              ObjectMap.toDirMap(map).get(objId).getInfo(
                  new ResultHandler<Directory.Info>(handlerId, resultCallbackCtxt)
              );
              return;
          }

          if (ObjectMap.isFileMap(map)) {
              ObjectMap.toFileMap(map).get(objId).getInfo(
                  new ResultHandler<File.Info>(handlerId, resultCallbackCtxt)
              );
          }
      }

      private void getDefDrive(JSONArray args, CallbackContext callbackContext) throws JSONException {
          Integer mapId = args.getInt(0);
          Integer objId = args.getInt(1);
          int handlerId = args.getInt(2);

          ObjectMap map = ObjectMap.acquire(mapId);
          if (ObjectMap.isClientMap(map)) {
              ObjectMap.toClientMap(map).get(objId).getDefaultDrive(
                  new ResultHandler<Drive>(handlerId, resultCallbackCtxt)
              );
              return;
          }
      }

      private void rootDirectory(JSONArray args, CallbackContext callbackContext) throws JSONException {
          Integer mapId = args.getInt(0);
          Integer objId = args.getInt(1);
          int handlerId = args.getInt(2);

          ObjectMap map = ObjectMap.acquire(mapId);
          if (ObjectMap.isDriveMap(map)) {
              ObjectMap.toDriveMap(map).get(objId).getRootDir(
                  new ResultHandler<Directory>(handlerId, resultCallbackCtxt)
              );
              return;
          }
      }

      private void createDirectory(JSONArray args, CallbackContext callbackContext) throws JSONException {
          Integer mapId = args.getInt(0);
          Integer objId = args.getInt(1);
          int handlerId = args.getInt(2);
          String  path  = args.getString(3);

          ObjectMap map = ObjectMap.acquire(mapId);
          if (ObjectMap.isDriveMap(map)) {
              ObjectMap.toDriveMap(map).get(objId).createDirectory(path,
                  new ResultHandler<Directory>(handlerId, resultCallbackCtxt)
              );
              return;
          }
          if (ObjectMap.isDirMap(map)) {
              ObjectMap.toDirMap(map).get(objId).createDirectory(path,
                  new ResultHandler<Directory>(handlerId, resultCallbackCtxt)
              );
              return;
          }
      }

      private void getDirectory(JSONArray args, CallbackContext callbackContext) throws JSONException {
          Integer mapId = args.getInt(0);
          Integer objId = args.getInt(1);
          int handlerId = args.getInt(2);
          String  path  = args.getString(3);

          ObjectMap map = ObjectMap.acquire(mapId);
          if (ObjectMap.isDriveMap(map)) {
              ObjectMap.toDriveMap(map).get(objId).getDirectory(path,
                  new ResultHandler<Directory>(handlerId, resultCallbackCtxt)
              );
              return;
          }
          if (ObjectMap.isDirMap(map)) {
              ObjectMap.toDirMap(map).get(objId).getDirectory(path,
                  new ResultHandler<Directory>(handlerId, resultCallbackCtxt)
              );
              return;
          }
      }

      private void createFile(JSONArray args, CallbackContext callbackContext) throws JSONException {
          Integer mapId = args.getInt(0);
          Integer objId = args.getInt(1);
          int handlerId = args.getInt(2);
          String  path  = args.getString(3);

          ObjectMap map = ObjectMap.acquire(mapId);
          if (ObjectMap.isDriveMap(map)) {
              ObjectMap.toDriveMap(map).get(objId).createFile(path,
                  new ResultHandler<File>(handlerId, resultCallbackCtxt)
              );
              return;
          }
          if (ObjectMap.isDirMap(map)) {
              ObjectMap.toDirMap(map).get(objId).createFile(path,
                  new ResultHandler<File>(handlerId, resultCallbackCtxt)
              );
              return;
          }
      }

      private void getFile(JSONArray args, CallbackContext callbackContext) throws JSONException {
          Integer mapId = args.getInt(0);
          Integer objId = args.getInt(1);
          int handlerId = args.getInt(2);
          String  path  = args.getString(3);

          ObjectMap map = ObjectMap.acquire(mapId);
          if (ObjectMap.isDriveMap(map)) {
              ObjectMap.toDriveMap(map).get(objId).getFile(path,
                  new ResultHandler<File>(handlerId, resultCallbackCtxt)
              );
              return;
          }
          if (ObjectMap.isDirMap(map)) {
              ObjectMap.toDirMap(map).get(objId).getFile(path,
                  new ResultHandler<File>(handlerId, resultCallbackCtxt)
              );
              return;
          }
      }

      private void getItemInfo(JSONArray args, CallbackContext callbackContext) throws JSONException {
          Integer mapId = args.getInt(0);
          Integer objId = args.getInt(1);
          int handlerId = args.getInt(2);
          String  path  = args.getString(3);

          ObjectMap map = ObjectMap.acquire(mapId);
          if (ObjectMap.isDriveMap(map)) {
              ObjectMap.toDriveMap(map).get(objId).getItemInfo(path,
                  new ResultHandler<ItemInfo>(handlerId, resultCallbackCtxt)
              );
              return;
          }
      }

      private void getChildren(JSONArray args, CallbackContext callbackContext) throws JSONException {
          Integer mapId = args.getInt(0);
          Integer objId = args.getInt(1);
          int handlerId = args.getInt(2);

          ObjectMap map = ObjectMap.acquire(mapId);
          if (ObjectMap.isDirMap(map)) {
              ObjectMap.toDirMap(map).get(objId).getChildren(
                  new ResultHandler<Children>(handlerId, resultCallbackCtxt)
              );
              return;
          }
      }

      private void moveTo(JSONArray args, CallbackContext callbackContext) throws JSONException {
          Integer mapId = args.getInt(0);
          Integer objId = args.getInt(1);
          int handlerId = args.getInt(2);
          String  path  = args.getString(3);

          ObjectMap map = ObjectMap.acquire(mapId);
          if (ObjectMap.isDirMap(map)) {
              ObjectMap.toDirMap(map).get(objId).moveTo(path,
                  new ResultHandler<Void>(handlerId, resultCallbackCtxt)
              );
              return;
          }
          if (ObjectMap.isFileMap(map)) {
              ObjectMap.toFileMap(map).get(objId).moveTo(path,
                  new ResultHandler<Void>(handlerId, resultCallbackCtxt)
              );
              return;
          }
      }

      private void copyTo(JSONArray args, CallbackContext callbackContext) throws JSONException {
          Integer mapId = args.getInt(0);
          Integer objId = args.getInt(1);
          int handlerId = args.getInt(2);
          String  path  = args.getString(3);

          ObjectMap map = ObjectMap.acquire(mapId);
          if (ObjectMap.isDirMap(map)) {
              ObjectMap.toDirMap(map).get(objId).copyTo(path,
                  new ResultHandler<Void>(handlerId, resultCallbackCtxt)
              );
              return;
          }
          if (ObjectMap.isFileMap(map)) {
              ObjectMap.toFileMap(map).get(objId).copyTo(path,
                  new ResultHandler<Void>(handlerId, resultCallbackCtxt)
              );
              return;
          }
      }

      private void deleteItem(JSONArray args, CallbackContext callbackContext) throws JSONException {
          Integer mapId = args.getInt(0);
          Integer objId = args.getInt(1);
          int handlerId = args.getInt(2);

          ObjectMap map = ObjectMap.acquire(mapId);
          if (ObjectMap.isDirMap(map)) {
              ObjectMap.toDirMap(map).get(objId).deleteItem(
                  new ResultHandler<Void>(handlerId, resultCallbackCtxt)
              );
              return;
          }
          if (ObjectMap.isFileMap(map)) {
              ObjectMap.toFileMap(map).get(objId).deleteItem(
                  new ResultHandler<Void>(handlerId, resultCallbackCtxt)
              );
              return;
          }
      }
  }
